### PR TITLE
`ssh`: Fix stealing of `'EXIT'` messages if a client is trapping exits

### LINKED ISSUE #8223: In OTP 26, Erlang's SSH client fails to connect if it receives an {'EXIT', Pid, normal} message from an unrelated process.
**Describe the bug**

If there’s an {‘EXIT’, _, normal} message in a process’ mailbox when calling ssh:connect(), the SSH library terminates without attempting to create a connection.

**To Reproduce**

Complete test case here:

```
%% Throwaway to debug SSH problems
%%
%% The Erlang SSH library doesn't seem to like old 'EXIT' messages left
%% lying around in the message queue. The mssage gets matched in
%% ssh_connection_handler:handshake/3; it looks like this behaviour was
%% added by 1963f9501
%%
%% I can reproduce the problem in otp_src_26.0.2
%% I cannot reproduce the problem in otp_src_23.1
%%
%%
-module(test_ssh_crash).

-export([works/0, fails/0]).

works() ->
    crypto:start(),
    ssh:start(),
    {ok, _SSH} = ssh:connect("localhost", 22,
			    [{user, "bogus"}, {password, "bogus"},
			     {silently_accept_hosts, true}]).

fails() ->
    self() ! {'EXIT', self(), normal},
    works().
```

**Expected behavior**

On  a machine with an SSH server, both works() and fails() should terminate with an authentication failure.

Actual behaviour is that fails() exits prematurely.

(And, if you replace the user/password with something valid, then works() will correctly connect to the SSH server, whereas fails() will not succeed with the connection.)

**Affected versions**

Seen in OTP 25 and OTP 26.

Bug possibly introduced in  1963f9501

**Additional context**

A workaround is to launch SSH from a process where you're sure you can't get any 'EXIT' messages.

### PR DESCRIPTION
**Pull request by Maria-12648430**
Fixes #8223.

I haven't written any tests for that yet. TBH, I don't know in which of the 20 suites I should put them :sweat_smile:

### CODE REVIEW

**Comment by u3s** [comment]
> I haven't written any tests for that yet. TBH, I don't know in which of the 20 suites I should put them 😅

I'm glad not to be the only one 😅
- in case of doubts, I would pick `ssh_connection_SUITE` as it seems a safe choice as everything in ssh is based on some _connections_ ;-)

**Comment by Maria-12648430** [comment]
Added two tests (one for `ssh:connect` and another for `ssh:daemon`) in the last commit.

File: lib/ssh/src/ssh_acceptor.erl [hunk 1]
```diff
@@ -164,17 +164,18 @@ handle_connection(_Address, _Port, {error,Error}, _Options, _Socket, _MaxSession
...
 handle_connection(Address, Port, _Peer, Options, Socket, _MaxSessions, _NumSessions, ParallelLogin)
   when ParallelLogin == false ->
-    handle_connection(Address, Port, Options, Socket);
+    handle_connection(undefined, Address, Port, Options, Socket);
 
 handle_connection(Address, Port, _Peer, Options, Socket, _MaxSessions, _NumSessions, ParallelLogin)
   when ParallelLogin == true ->
+    Parent = self(),
     Ref = make_ref(),
     Pid = spawn_link(
```
Comment (IngelaAndin) [inline lib/ssh/src/ssh_acceptor.erl:173]: Spawning this process here that is not part of the supervisor tree  and linking it to the user process I think is problematic.

File: lib/ssh/src/ssh_connection_handler.erl [hunk 2]
```diff
@@ -121,11 +121,11 @@ start_link(Role, _Address=#address{}, Id, Socket, Options) ->
...
 
 
-takeover(ConnPid, client, Socket, Options) ->
+takeover(ParentPid, ConnPid, client, Socket, Options) ->
     group_leader(group_leader(), ConnPid),
-    takeover(ConnPid, common, Socket, Options);
+    takeover(ParentPid, ConnPid, common, Socket, Options);
 
-takeover(ConnPid, _, Socket, Options) ->
+takeover(ParentPid, ConnPid, _, Socket, Options) ->
```
Comment (IngelaAndin) [inline lib/ssh/src/ssh_connection_handler.erl:128]: Here we will be changing  controlling process a second time. I think we should write the code in such a way that we only need to do it once. The acceptor process that should be  in the supervisor tree should create the socket and start  the appropriate subsystem dynamic supervisor tree and retrieve the pid of the connection_handler procsee and transfer the ownership once.

**Comment by IngelaAndin** [comment]
@Maria-12648430  I add some thoughts on how I think it should work.

**Comment by IngelaAndin** [comment]
Also I think it should work to have several accepting processes, first one will win and the others might get the next connection attempt.

**Comment by Maria-12648430** [comment]
> @Maria-12648430 I add some thoughts on how I think it should work.

Thanks @IngelaAndin, I will take a closer look next week, I'm currently on a short vacation 🙂

**Comment by Maria-12648430** [comment]
@IngelaAndin your suggestions make sense to me and I think we should go that way. However, this makes it a larger refactoring (vs a rather simple bugfix) which will take some time to work out.

**Comment by IngelaAndin** [comment]
@Maria-12648430 no problem Maria, I think it is important to fix things the "tm correct way" because otherwise it is mainly a question of time until some new problem pops up. I believe in fixing the  problem not only shutting the symptoms up. We very much apricate you taking the time, as we alas not always have enough resources to fix everything as swiftly as we may desire.
It was an unfortunate mistake I think this design, that probably slipped through as we might have missed to review some code, but code reviewing activities are also a tradeoff between other things that needs doing, so we just have to deal with it.

**Comment by u3s** [comment]
> Also I think it should work to have several accepting processes, first one will win and the others might get the next connection attempt.

Do you mean to have that in scenario with parallel_login == true or parallel_login == false?

**Comment by u3s** [comment]
I agree with suggestions from Ingela.
When browsing code I also found code line which is probably redundant.

`user_pid` option is added twice - _probably safe to remove (tm)_ ;-) if turns out related to this work. 
`ssh_acceptor:handle_connection/4` invokes `?PUT_INTERNAL_OPT([{user_pid,` which is then repeated in `ssh_system_sup:start_subsystem/4`

I think instruction in ssh_acceptor could be removed.

File: lib/ssh/test/ssh_connection_SUITE.erl [hunk 3]
```diff
@@ -1174,6 +1178,59 @@ do_start_shell_exec_fun(Fun, Command, Expect, ExpectType, Config) ->
...
+
+    %% Ensure the EXIT message is still there
+    receive
+        ExitMsg -> ok
+    after 0 ->
+        ct:fail("No EXIT message")
+    end.
+
+%%--------------------------------------------------------------------
+%% Issue #8223
```
Comment (u3s) [inline lib/ssh/test/ssh_connection_SUITE.erl:1213]: if you write `GH-8223` it will be clear you're referring to github issue.

**Comment by Maria-12648430** [comment]
This is pretty hard stuff :cold_sweat: Everything happens all over the place and gets pushed around to everywhere else, or so it seems 🤪 This will take a while to sort out and not break anything.

**Comment by IngelaAndin** [comment]
> > Also I think it should work to have several accepting processes, first one will win and the others might get the next connection attempt.
> 
> Do you mean to have that in scenario with parallel_login == true or parallel_login == false?
I guess it makes little sense for parallel_login false! I think this option makes little sense at all. I think you could always have parallel login and put the limit to one to get false behavior.

**Comment by Maria-12648430** [comment]
Ok, with help from @juhlig, this is what we came up with. Let us know what you think.

We put another simple_one_for_one supervisor `ssh_acceptor_subsup` between `ssh_acceptor_sup` and `ssh_acceptor`. It sets up the listening socket for a daemon (this was formerly in `ssh_acceptor`). The actual acceptor processes are children of this supervisor.

We also introduced a new value for the `parallel_login` option in addition to `true` and `false` which can be a positive integer.

The overall behavior is as follows, depending on the value given for `parallel_login`:
* If `parallel_login` is a number `N` (`false` is synonymous to `N=1`), that number of `ssh_acceptor` children are started. Each of those acceptors runs in a loop that waits for a connection, handles it, then wraps around to waiting for the next connection. This way, up to `N` login attempts can be handled simultaneously; if there are more than `N` simultaneous login attempts, they have to wait.
* If `parallel_login` is `true` (in the sense of an unlimited number of simultaneous login attempts), only 1 `ssh_acceptor` child is started initially. As soon as this acceptor gets a connection, it tells the supervisor to start another acceptor before handling the connection and finally exiting (instead of looping around to wait for another connection). The freshly started acceptor will behave in the same way.

The outlined behavior introduces no backwards incompatibilities. `parallel_login` values `false` and `true` remain with the same outward behavior, and the newly introduced value `N` provides a good "middle ground" between sequential accepting and unlimited all-out parallelism.

Moreover, the approach solves the issue (#8223) that gave raise to this PR, and as far as we can tell incorporate the suggestions made before.

There is one failing test, though: `ssh_protocol_SUITE:client_close_after_hello/1` (there may be more, some are being skipped on my machine, don't know about those). I don't really grasp what this one is supposed to be testing, and if it still applies with the changes made. I would be grateful if you could have a look at it.

File: lib/ssh/src/ssh_acceptor_subsup.erl [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    restart => transient}],
+    {ok, {SupFlags,ChildSpecs}}.
+
+%%%=========================================================================
+%%%  Internal functions
+%%%=========================================================================
+
+listen(Port, Options) ->
+    {_, Callback, _} = ?GET_OPT(transport, Options),
+    SockOpts = ?GET_OPT(socket_options, Options) ++ [{active, false}, {reuseaddr, true}],
```
Comment (Maria-12648430) [inline lib/ssh/src/ssh_acceptor_subsup.erl:120]: This was `[{active, false}, {reuseaddr,true} | ?GET_OPT(socket_options, Options)]` in `ssh_acceptor`, which is wrong. For one, it would prevent the usage of option `{inet_backend, socket}` since this option must be the first in the list. For another, in options given to `gen_tcp` or `ssl`, the _last_, option wins, such that if for example `{active, true}` was in the list of given socket options, it would be set to `{active, true}` despite the tacked-on `{active, false}`.

**Comment by u3s** [comment]
> @IngelaAndin @u3s Ok, with help from @juhlig, this is what we came up with. Let us know what you think.

Thanks a lot for work. We planned code review and will get back.

File: lib/ssh/src/ssh_acceptor.erl [hunk 5]
```diff
@@ -42,24 +41,10 @@
 %% Internal application API
 %%====================================================================
 %% Supposed to be called in a child-spec of the ssh_acceptor_sup
-start_link(SystemSup, Address, Options) ->
-    proc_lib:start_link(?MODULE, acceptor_init, [self(),SystemSup,Address,Options]).
+start_link(SystemSup, LSock, Address, Options) ->
+    proc_lib:start_link(?MODULE, acceptor_init, [self(), SystemSup, LSock, Address, Options]).
```
Comment (IngelaAndin) [inline lib/ssh/src/ssh_acceptor.erl:45]: Here we can use proc_lib:spawn_link or proc_lib:sapwn_opt with link option, instead of start_link to avoid having to use proc_lib:init_ack as all processes under the simple_one_for_one are independent of each other and do not have to wait for each others init functions to complete.

**Comment by IngelaAndin** [comment]
I think the test looks like a white-box test, and probably it trying to test that there is no process leak if the client gives up the connection after sending its hello message. Probably it should be checked differently as you have changed the supervisor tree.  

I like the way that you describe in the documentation what should happen. I have not full understood the code yet. I think I need to look at it as whole the diff is not enough.

**Comment by Maria-12648430** [comment]
@IngelaAndin looking forward to your review :slightly_smiling_face: I'll see what I can do about that test next week.

Btw, I got a notification by email about a comment from you regarding `proc_lib:spawn_link` vs `proc_lib:start_link`+`proc_lib:init_ack` in `ssh_acceptor`, but strangely I can't find it here in Github :thinking:

**Comment by IngelaAndin** [comment]
@Maria-12648430  yes it was because I thought maybe my comment was premature, so I deleted it for the moment. It is true that processes under a simple_one_for_one supervisor would benefit from using proc_lib:spawn_link and then use <behaviour>:enter_loop instead of  proc_lib:start_link as they are independent of each other and do not need to wait for each others init, so we can skip send and receiving of the ack message. Question is, if I had put this comment in the right place. So now I just made a general comment instead.

File: lib/ssh/src/ssh_acceptor_subsup.erl [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                       false -> 1;
+                       N -> N
+                   end,
+    {Sock, Opts} = case ?GET_INTERNAL_OPT(lsocket, Options, undefined) of
+        {LSock, SockOwner} ->
+            %% A listening socket (or fd option) was provided in the ssh:daemon call
+            case inet:sockname(LSock) of
+                {ok, {_, Port}} ->
+                    %% A usable, open LSock
+                    spawn_link(fun() ->
```
Comment (IngelaAndin) [inline lib/ssh/src/ssh_acceptor_subsup.erl:69]: I realize this is probably the fault of the original design, but transferring ownership of the listen socket from ssh user process context feels like a bad idea. Why not make the listen call in a dedicated acceptor process? In current solution that would be in the ssh_acceptor_subsup process.

File: lib/ssh/src/ssh_acceptor_subsup.erl [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                              unlink(Self)
+                          end),
+                    {LSock, Options};
+
+                {error, _Error} ->
+                    %% Not open, a restart
+                    %% Allow gen_tcp:listen to fail 4 times if eaddrinuse (It is a bug fix):
+                    case try_listen(Port, Options, 4) of
+                        {ok, NewLSock} ->
+                            spawn_link(fun() ->
```
Comment (IngelaAndin) [inline lib/ssh/src/ssh_acceptor_subsup.erl:82]: Why do we do spawn_link and unlink? We could do just spawn if we do not want the link?

Comment (Maria-12648430) [reply lib/ssh/src/ssh_acceptor_subsup.erl:82]: Oh, that's a relic :sweat_smile: Initially, I had more stuff going on in this function, and the `spawn_link` was there to stop the spawned process in case the supervisor exits while it was doing that. The `unlink` was there to prevent the supervisor receiving and complaining about an unexpected `'EXIT'` message when the spawned process exits normally.

tl;dr: I'll change it as you suggest, also on lines 96 and 98.

File: lib/ssh/src/ssh_acceptor_subsup.erl (hunk 6 above)
Comment (Maria-12648430) [reply lib/ssh/src/ssh_acceptor_subsup.erl:69]: I was thinking the same thing, but see here: https://github.com/erlang/otp/blob/9859b7281e447681a38da9f34292a7825f437de7/lib/ssh/src/ssh.erl#L599-L613

File: lib/ssh/src/ssh_acceptor_subsup.erl (hunk 7 above)
Comment (Maria-12648430) [reply lib/ssh/src/ssh_acceptor_subsup.erl]: done :+1:

File: lib/ssh/src/ssh_acceptor_subsup.erl (hunk 6 above)
Comment (IngelaAndin) [reply lib/ssh/src/ssh_acceptor_subsup.erl:69]: Hum, well maybe we need it then. I am still not fond of making an explicit receive in the users process like that without more control like in gen_server:call for instance.

Comment (Maria-12648430) [reply lib/ssh/src/ssh_acceptor_subsup.erl:69]: I'm not a huge fan of that, either... Maybe a greater refactoring attempt is in order then, but I think that should involve more planning.

Comment (IngelaAndin) [reply lib/ssh/src/ssh_acceptor_subsup.erl:69]: Maybe we could go with a "good enough bug fix" and then make a refactor later in new master after OTP-27 (as it is really soon)

**Comment by Maria-12648430** [comment]
We are still stuck on this one test. It just doesn't add up.

Before the structural changes in this PR, when parallel login was enabled, the acceptor would spawn_link a new process to do the authentication and handshake, which would trap exits. If the acceptor died during this, it would notice the exit via the link and stop the connection handler process again. This is what it looks like the test covers: it sends an exit to the acceptor (which is _not_ trapping exits), then checks that there are no processes in handshake and no connection handler processes.

Notably, when parallel login was disabled, the implementation would _not_ stop a connection handler process hanging in handshake. The acceptor process was doing the authentication and handshake by itself (ie, not in a spawned process), and the death of the acceptor would just stop the process. The handshake might time out, that would be kinda ok, but if it succeeded later there would be an orphaned connection handler process.

Notably again, this can also happen when a user starts a client via `connect`. If the client dies when the connection is in handshake, the connection handler process will not be stopped.

With the changes in this PR, there are now multiple acceptors handling authorization and handshake when parallel login is enabled, each handling authentication and handshake by itself like the single acceptor in the old implementation with parallel login disabled. When the acceptor receives an exit and dies, the connection handler is orphaned the same way.

The acceptor process could be changed to trap exits and shut down connections stuck in handshake, but as the selective receive in the handshake does not have a clause, this could only be done after the handshake eventually succeeded or timed out. Adding another clause is also not a good option since the same function is also used in the user process (which gave raise to the initial issue). So it looks like this cannot be solved without more structural changes, that is, another layer of separate processes independent from the acceptor and user processes. But as I already said above, this gets out of hand and turns into a bigger refactoring which requires better planning. What to do? :cold_sweat: (@IngelaAndin)

**Comment by juhlig** [comment]
> What to do?

> Maybe we could go with a "good enough bug fix" and then make a refactor later in new master after OTP-27 (as it is really soon)

I guess the only "good enough bugfix" that we can reasonably aim for now, without going into large-scale restructuring, would be what we had in the beginning: Revert the structural changes we have now (bring them back later), instead keep the spawn_linking of a process by the acceptor, but nail the receive of the `'EXIT'` in handshake down to the acceptor pid :man_shrugging:

File: lib/ssh/src/ssh_acceptor_subsup.erl (hunk 6 above)
Comment (IngelaAndin) [reply lib/ssh/src/ssh_acceptor_subsup.erl:69]: I think that we should be able to move away from the above approach. Th concept of fd is provided is inherited from httpd_sup and  we should be able to handle it like that, that is port is also specified but not opened as it is preopend.
And the case of 0 is the port is only for testing and I think we should rather report the the chosen port bac to the caller
then the way it is done now.

**Comment by IngelaAndin** [comment]
@juhlig @Maria-12648430 We discussed it at meeting today and we feel that we will try avoiding making a good enough bug-fix   as it might introduce new strangeness. We rather like to continue working on a rewrite in the OTP-27 track that possible could be backported later when it has proven itself.

**Comment by IngelaAndin** [comment]
@Maria-12648430 sure we can close it and we are hoping you replace it with a new and even better one :)  We really apricate your help with this. We are kind of swamped right now in the security protocol area so to speak.

**Comment by Maria-12648430** [comment]
Ok then, I'll dig into the code to better understand the status quo, see what I can do, and come back when I have something to show :smiley:

File: lib/ssh/test/ssh_connection_SUITE.erl [hunk 8]
```diff
@@ -1175,6 +1179,59 @@ do_start_shell_exec_fun(Fun, Command, Expect, ExpectType, Config) ->
...
+    %% Ensure the EXIT message is still there
+    receive
+        ExitMsg -> ok
+    after 0 ->
+        ct:fail("No EXIT message")
+    end.
+
+%%--------------------------------------------------------------------
+%% Issue GH-8223
+trap_exit_daemon(Config) when is_list(Config) ->
```
Comment (u3s) [inline lib/ssh/test/ssh_connection_SUITE.erl:1214]: Have you seen `trap_exit_daemon` testcase failing on clean master branch? In my execution it always succeeds.
By reading execution traces I don't thin 'EXIT' issue occurs for `ssh:daemon` calls

Comment (Maria-12648430) [reply lib/ssh/test/ssh_connection_SUITE.erl:1214]: I don't really remember, but I don't think it did, and we probably added this test just to be sure 🤷🏼‍♀️ I'm on vacation ATM, reading code on my tiny mobile is difficult, the codebase is complex... tl;dr, reasoning about whether or not this _could_ happen is pretty hard for me right now 😅
